File: LeetCode/2265.Count_Nodes_Equal_to_Average_of_Subtree/2265.Count_Nodes_Equal_to_Average_of_Subtree.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        self.res = 0

        # dfs input node sum count
        # return (sum, count)
        def dfs(node):
            if not node:
                return (0, 0)

            ls, lc = dfs(node.left)
            rs, rc = dfs(node.right)

            count = lc + rc + 1
            acc = ls + rs + node.val

            if node.val == acc // count:
                self.res += 1

            # print(f"acc:{acc} count:{count}")
            return (acc, count)

        dfs(root)
        return self.res

File: LeetCode/2265.Count_Nodes_Equal_to_Average_of_Subtree/2265.Count_Nodes_Equal_to_Average_of_Subtree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func averageOfSubtree(_ root: TreeNode?) -> Int {
        // the num of nodes, val == average of substree
        var res = 0

        // sum of tree, the num of node
        func dfs(_ node: TreeNode?) -> (Int, Int) {
            guard let node = node else { return (0, 0) }

            let (lSum, lCnt) = dfs(node.left)
            let (rSum, rCnt) = dfs(node.right)

            let sum = lSum + rSum + node.val
            let cnt = lCnt + rCnt + 1

            if sum / cnt == node.val {
                res += 1
            }

            return (sum, cnt)
        }

        dfs(root)

        return res
    }
}
